tests/extra/job_management/test_thread_worker.py

```diff
@@ -0,0 +1,204 @@
+    """Job id generator that extracts the year from the process graph"""
+    try:
+        (year,) = (n["arguments"]["year"] for n in process_graph.values())
+        return f"job-{year}"
+    except Exception:
+        pass
+
+
+@pytest.fixture
+def dummy_backend_foo(requests_mock) -> DummyBackend:
+    dummy = DummyBackend.at_url("https://foo.test", requests_mock=requests_mock)
+    dummy.setup_simple_job_status_flow(queued=3, running=5)
+    dummy.job_id_generator = _job_id_from_year
+    return dummy
+
+
+@pytest.fixture
+def dummy_backend_bar(requests_mock) -> DummyBackend:
+    dummy = DummyBackend.at_url("https://bar.test", requests_mock=requests_mock)
+    dummy.setup_simple_job_status_flow(queued=5, running=8)
+    dummy.job_id_generator = _job_id_from_year
+    return dummy
+
+
+@pytest.fixture
+def sleep_mock():
+    with mock.patch("time.sleep") as sleep:
+        yield sleep
+
+
+class TestJobManagerWorkerThread:
+
+    @pytest.fixture
+    def job_manager_root_dir(self, tmp_path):
+        return tmp_path / "job_mgr_root"
+
+    @pytest.fixture
+    def job_manager(self, job_manager_root_dir, dummy_backend_foo, dummy_backend_bar):
+        manager = MultiBackendJobManager(root_dir=job_manager_root_dir)
+        manager.add_backend("foo", connection=dummy_backend_foo.connection)
+        manager.add_backend("bar", connection=dummy_backend_bar.connection)
+        return manager
+
+    @staticmethod
+    def _create_year_job(row, connection, **kwargs):
+        year = int(row["year"])
+        pg = {"yearify": {"process_id": "yearify", "arguments": {"year": year}, "result": True}}
+        return connection.create_job(pg)
+
+    @pytest.fixture
+    def fresh_worker(self):
+        work_queue = queue.Queue()
+        result_queue = queue.Queue()
+        worker = _JobManagerWorkerThread(work_queue, result_queue)
+        yield worker
+        if worker.is_alive():
+            worker.stop_event.set()
+            worker.join(timeout=1)
+
+    def enqueue_and_wait(self, worker, work_item, timeout=0.2):
+        worker.work_queue.put(work_item)
+        worker.start()
+        sleep(timeout)
+        worker.stop_event.set()
+        worker.join(timeout=1)
+
+    def test_run_jobs_collects_worker_results(self, tmp_path, job_manager, sleep_mock):
+        df = pd.DataFrame({"year": [2018], "geometry": ["POINT (1 2)"]})
+        job_db_path = tmp_path / "jobs.csv"
+        job_db = CsvJobDatabase(job_db_path).initialize_from_df(df)
+
+        captured_results = []
+        original_process_result_queue = job_manager._process_result_queue
+
+        def spy_process_result_queue(stats=None):
+            try:
+                work_result = job_manager._result_queue.get_nowait()
+                captured_results.append(work_result)
+            except queue.Empty:
+                pass
+            original_process_result_queue(stats)
+
+        job_manager._process_result_queue = spy_process_result_queue
```

Comment:
you are not only spying here, but also stealing a result from the result queue. This might cause inconsistencies, or a test stuck in infinite waiting loop